Mathlib/Data/Set/Basic.lean

```diff
@@ -2168,4 +2168,28 @@ end Disjoint
 @[simp] theorem Prop.compl_singleton (p : Prop) : ({p}ᶜ : Set Prop) = {¬p} :=
   ext fun q ↦ by simpa [@Iff.comm q] using not_iff
 
+namespace Equiv
+
+/-- Given a predicate `p : α → Prop`, produces an equivalence between
+  `Set {a : α // p a}` and `{s : Set α // ∀ a ∈ s, p a}`.
+  -/
+protected def setSubtypeCongr {α : Type*} (p : α → Prop) :
+    Set {a : α // p a} ≃ {s : Set α // ∀ a ∈ s, p a} where
+  toFun s := ⟨fun a ↦ ∃ h : p a, s ⟨a, h⟩, fun _ h ↦ h.1⟩
+  invFun s a := s.val a
+  left_inv s := by ext a; exact ⟨fun h ↦ h.2, fun h ↦ ⟨a.property, h⟩⟩
+  right_inv s := by ext; exact ⟨fun h ↦ h.2, fun h ↦ ⟨s.property _ h, h⟩⟩
+
+@[simp]
+protected lemma setSubtypeCongr_apply_coe (p : α → Prop) (s : Set {a // p a}) :
+    (Equiv.setSubtypeCongr p) s = ⟨fun a ↦ ∃ h : p a, s ⟨a, h⟩, fun _ h ↦ h.1⟩ :=
```

Comment:
You're breaking the API barrier for `Set` here. You should never apply a set to an argument.